src/scenic/core/shapes.py

```diff
@@ -0,0 +1,151 @@
+        return region_distance > 0
+
+    @property
+    @abstractmethod
+    def mesh(self):
+        pass
+
+    @property
+    @abstractmethod
+    def isConvex(self):
+        pass
+
+###################################################################################################
+# 3D Shape Classes
+###################################################################################################
+
+class MeshShape(Shape):
+    """ A Shape subclass defined by a Trimesh object.
+
+    Args:
+        mesh: A trimesh.Trimesh mesh object.
+        dimensions: The raw (before scaling) dimensions of the shape. If dimensions
+          and scale are both specified the dimensions are first set by dimensions, and then
+          scaled by scale.
+        scale: Scales all the dimensions of the shape by a multiplicative factor.
+          If dimensions and scale are both specified the dimensions are first set by dimensions,
+          and then scaled by scale.
+        initial_rotation: A 3-tuple containing the yaw, pitch, and roll respectively to apply when loading
+          the mesh. Note the initial_rotation must be fixed.
+    """
+    def __init__(self, mesh, dimensions=None, scale=1, initial_rotation=None):
+        # Ensure the mesh is watertight so volume is well defined
+        if not mesh.is_watertight:
+            raise ValueError("A MeshShape cannot be defined with a mesh that does not have a well defined volume.")
+
+        # Copy mesh and center vertices around origin
+        self._mesh = mesh.copy()
+        self._mesh.vertices -= self._mesh.bounding_box.center_mass
+        self._isConvex = self._mesh.is_convex
+
+        # If rotation is provided, apply rotation
+        if initial_rotation is not None:
+            rotation = Orientation.fromEuler(*initial_rotation)
+            rotation_matrix = quaternion_matrix((rotation.w, rotation.x, rotation.y, rotation.z))
+            self._mesh.apply_transform(rotation_matrix)
+
+        # If dimensions are not specified, infer them.
+        if dimensions is None:
+            dimensions = list(self._mesh.extents)
+
+        # Scale mesh to unit size
+        scale_vals = self._mesh.extents / numpy.array([1,1,1])
+        scale_matrix = numpy.eye(4)
+        scale_matrix[:3, :3] /= scale_vals
+        self._mesh.apply_transform(scale_matrix)
+
+        # Report samplables
+        super().__init__(dimensions, scale)
+
+    @classmethod
+    def fromFile(cls, path, filetype=None, compressed=None, **kwargs):
+        """Load a shape from a file, attempting to infer filetype and compression.
+
+        For example: "foo.obj.bz2" is assumed to be a compressed .obj file.
+        "foo.obj" is assumed to be an uncompressed .obj file. "foo" is an
+        unknown filetype, so unless a filetype is provided an exception will be raised.
+
+        Args:
+            path: Path to the file to import
+            filetype: Filetype of file to be imported. This will be inferred if not provided
```

Comment:
Not clear what the acceptable types or values for this parameter are. It may be enough to just state the type and link to the trimesh docs for the accepted formats (if we place no further restrictions). Since those docs also use Sphinx I would add trimesh to `intersphinx` so you can do cross-references like \`trimesh.Trimesh\` throughout.